src/Microsoft.SqlTools.ServiceLayer/Agent/AgentService.cs

```diff
@@ -1211,6 +1219,215 @@ public static XmlDocument CreateJobXmlDocument(string svrName, string jobName)
+                }
+            });
+        }
+
+        internal async Task HandleAgentNotebookHistoryRequest(AgentNotebookHistoryParams parameters, RequestContext<AgentNotebookHistoryResult> requestContext)
+        {
+            await Task.Run(async () =>
+            {
+                try
+                {
+                    var result = new AgentNotebookHistoryResult();
+                    ConnectionInfo connInfo;
+                    ConnectionServiceInstance.TryFindConnection(
+                        parameters.OwnerUri,
+                        out connInfo);
+                    if (connInfo != null)
+                    {
+                        string getNotebookHistoryQueryString = 
+                            "USE "+ parameters.TargetDatabase + ";" +
+                            "SELECT * FROM notebooks.nb_materialized " +
+                            "WHERE JOB_ID = '" + parameters.JobId + "'";
+
+                        DataSet NotebookHistoryDataset = executeQuery(connInfo, getNotebookHistoryQueryString);
+                        CDataContainer dataContainer = CDataContainer.CreateDataContainer(connInfo, databaseExists: true);
+                        var jobServer = dataContainer.Server.JobServer;
+                        var jobs = jobServer.Jobs;
+                        Tuple<SqlConnectionInfo, DataTable, ServerConnection> tuple = CreateSqlConnection(connInfo, parameters.JobId);
+                        SqlConnectionInfo sqlConnInfo = tuple.Item1;
+                        DataTable dt = tuple.Item2;
+                        ServerConnection connection = tuple.Item3;
+
+                        var agentNotebookHistories = new List<AgentNotebookHistoryInfo>();
+                        JobStepCollection steps = jobs[parameters.JobName].JobSteps;
+                        var jobSteps = new List<AgentJobStepInfo>();
+                        foreach (JobStep step in steps)
+                        {
+                            jobSteps.Add(AgentUtilities.ConvertToAgentJobStepInfo(step, parameters.JobId, parameters.JobName));
+                        }
+                        result.Steps = jobSteps.ToArray();
+
+                        // Add schedules to the job if any
+                        JobScheduleCollection schedules = jobs[parameters.JobName].JobSchedules;
+                        var jobSchedules = new List<AgentScheduleInfo>();
+                        foreach (JobSchedule schedule in schedules)
+                        {
+                            jobSchedules.Add(AgentUtilities.ConvertToAgentScheduleInfo(schedule));
+                        }
+                        result.Schedules = jobSchedules.ToArray();
+                        // Add histories
+                        int count = dt.Rows.Count;
+                        List<AgentNotebookHistoryInfo> notebookHistories = new List<AgentNotebookHistoryInfo>();
+                        if (count > 0)
+                        {
```

Comment:
If count is <= 0 you never send a response and thus will time out on the ADS side. You should send an error in that case.